File: web/sdesign-web/src/app/models/eqConfig.ts
import { ShowConfig } from "./showConfig";

export interface EQConfig {
  bypass: boolean;
  high_pass_enable: boolean;
  high_pass_gain: number;
  mid_enable: boolean;
  mid_gain: number;
  low_pass_enable: boolean;
  low_pass_gain: number;
}

export function SetEQConfigValue(cfg: ShowConfig, path: string[], value: string): ShowConfig {
  switch (path[3]) {
    case "bypass":
      cfg.channel_cfgs[+path[1]].eq_cfg.bypass = value.toLowerCase() == 'true';
      break;
    case "high_pass_enable":
      cfg.channel_cfgs[+path[1]].eq_cfg.high_pass_enable = value.toLowerCase() == 'true';
      break;
    case "high_pass_gain":
      cfg.channel_cfgs[+path[1]].eq_cfg.high_pass_gain = +value;
      break;
    case "mid_enable":
      cfg.channel_cfgs[+path[1]].eq_cfg.mid_enable = value.toLowerCase() == 'true';
      break;
    case "mid_gain":
      cfg.channel_cfgs[+path[1]].eq_cfg.mid_gain = +value;
      break;
    case "low_pass_enable":
      cfg.channel_cfgs[+path[1]].eq_cfg.low_pass_enable = value.toLowerCase() == 'true';
      break;
    case "low_pass_gain":
      cfg.channel_cfgs[+path[1]].eq_cfg.low_pass_gain = +value;
      break;
  }
  return cfg;
}

File: backend/config/eqConfig.go
package config

import (
	"fmt"
	"strconv"
)

type EQConfig struct {
	Bypass         bool    `json:"bypass"`
	HighPassEnable bool    `json:"high_pass_enable"`
	HighPassGain   float64 `json:"high_pass_gain"`
	MidEnable      bool    `json:"mid_enable"`
	MidGain        float64 `json:"mid_gain"`
	LowPassEnable  bool    `json:"low_pass_enable"`
	LowPassGain    float64 `json:"low_pass_gain"`
}

func NewEQConfig() *EQConfig {
	c := EQConfig{
		Bypass:         true,
		HighPassEnable: false,
		HighPassGain:   0,
		MidEnable:      false,
		MidGain:        0,
		LowPassEnable:  false,
		LowPassGain:    0,
	}
	return &c
}

func (c *EQConfig) GetValue(path []string) (string, error) {
	switch path[0] {
	case "bypass":
		return strconv.FormatBool(c.Bypass), nil
	case "high_pass_enable":
		return strconv.FormatBool(c.HighPassEnable), nil
	case "high_pass_gain":
		return strconv.FormatFloat(c.HighPassGain, 'f', -1, 64), nil
	case "mid_enable":
		return strconv.FormatBool(c.MidEnable), nil
	case "mid_gain":
		return strconv.FormatFloat(c.MidGain, 'f', -1, 64), nil
	case "low_pass_enable":
		return strconv.FormatBool(c.LowPassEnable), nil
	case "low_pass_gain":
		return strconv.FormatFloat(c.LowPassGain, 'f', -1, 64), nil
	default:
		return "", fmt.Errorf("encountered unexpected path variable %s", path[0])
	}
}

func (c *EQConfig) SetValue(path []string, value string) error {
	switch path[0] {
	case "bypass":
		val, err := strconv.ParseBool(value)
		if err != nil {
			return err
		}
		c.Bypass = val
		return nil
	case "high_pass_enable":
		val, err := strconv.ParseBool(value)
		if err != nil {
			return err
		}
		c.HighPassEnable = val
		return nil
	case "high_pass_gain":
		val, err := strconv.ParseFloat(value, 64)
		if err != nil {
			return err
		}
		c.HighPassGain = val
		return nil
	case "mid_enable":
		val, err := strconv.ParseBool(value)
		if err != nil {
			return err
		}
		c.MidEnable = val
		return nil
	case "mid_gain":
		val, err := strconv.ParseFloat(value, 64)
		if err != nil {
			return err
		}
		c.MidGain = val
		return nil
	case "low_pass_enable":
		val, err := strconv.ParseBool(value)
		if err != nil {
			return err
		}
		c.LowPassEnable = val
		return nil
	case "low_pass_gain":
		val, err := strconv.ParseFloat(value, 64)
		if err != nil {
			return err
		}
		c.LowPassGain = val
		return nil
	default:
		return fmt.Errorf("encountered unexpected path variable %s", path[0])
	}
}
